File: array/leetcode_349.java
package array;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class leetcode_349 {
    public static void main(String[] args) {
        leetcode_349 lc = new leetcode_349();
        int[] nums1 = {1, 2, 2, 1};
        int[] nums2 = {2, 2};
        System.out.println(Arrays.toString(lc.intersection(nums1, nums2)));
    }

    public int[] intersection(int[] nums1, int[] nums2) {
        Set<Integer> set = new HashSet<>();
        Arrays.sort(nums2);

        for (Integer num : nums1) {
            if (binarySearch(nums2, num)) set.add(num);
        }

        return set.stream().mapToInt(i -> i).toArray();
    }

    private boolean binarySearch(int[] nums, int target) {
        int low = 0, high = nums.length - 1;

        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (nums[mid] == target) return true;
            if (nums[mid] > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return false;
    }
}

File: array/LeetCode_349.kt
package array

class LeetCode_349 {

    fun intersection(nums1: IntArray, nums2: IntArray): IntArray {
        val numSet = hashSetOf<Int>()

        nums2.sort()

        nums1.forEach { num ->
            if (isContains(nums2, num)) {
                numSet.add(num)
            }
        }

        return numSet.toIntArray()
    }

    private fun isContains(nums: IntArray, target: Int): Boolean {
        var low = 0
        var high = nums.lastIndex

        while (low <= high) {
            val mid = low + (high - low) / 2

            if (nums[mid] == target) {
                return true
            } else if (nums[mid] > target) {
                high = mid - 1
            } else {
                low = mid + 1
            }
        }

        return false
    }
}

fun main() {
    // Input
    val nums1 = intArrayOf(1, 2, 2, 1)
    val nums2 = intArrayOf(2, 2)
    // Output
    LeetCode_349().intersection(nums1, nums2).run {
        println(this.joinToString())
        require(
            intArrayOf(2).contentEquals(this)
        )
    }
}
